copy-tex contrib module

edemaine opened the pull request:
Solution to #645.  We wouldn't want this in base KaTeX, as it globally modifies copy/paste behavior, and a user might want to tweak it further.  But it's usable as-is for anyone not wanting to otherwise modify the copy/paste behavior, and a useful starting point in general.

From the README:

> This extension modifes the copy/paste behavior in any browser supporting the [Clipboard API](https://developer.mozilla.org/en-US/docs/Web/API/ClipboardEvent) so that, when selecting and copying whole KaTeX-rendered elements, the text content of the resulting clipboard renders KaTeX elements as their LaTeX source surrounded by specified delimiters.  (The HTML content remains unchanged.) The default delimiters are `$...$` for inline math and `$$...$$` for display math, but you can easy switch them to e.g. `\(...\)` and `\[...\]` by modifying `copyDelimiters` in [the source code](copy-tex.js).

[Demo on jsfiddle](https://jsfiddle.net/edemaine/hazzrd1d/) (slightly older version of the code)

kevinbarabash commented:
I've got stung a couple of times pushing changes with lint in it.  Maybe we could run lint as a pre-commit or pre-push hook.

kevinbarabash commented:
http://caniuse.com/#search=clipboard

<img width="1213" alt="screen shot 2017-08-21 at 2 33 36 pm" src="https://user-images.githubusercontent.com/1044413/29532984-c27319c8-867d-11e7-86c6-c3e55943b6c9.png">

edemaine commented:
Now fully tested and working in Chrome (Windows and Android), Firefox (Windows), and Edge.  Can someone test Safari?

I found that Firefox had trouble with the modern ES I used (particularly [`replaceWith`](https://developer.mozilla.org/en-US/docs/Web/API/ChildNode/replaceWith), and in some older versions, `Array.forEach`), so I implemented some workarounds.

I found that Edge didn't like being given a DocumentFragment, so I converted it to HTML.  Also, given two versions of the copy/paste buffer, it just kept the latter one.  ([This is a bug in Edge](https://developer.microsoft.com/en-us/microsoft-edge/platform/status/clipboardapi/).)  So I now include the HTML before the plain text, so the latter gets used for Edge.  (This actually also makes the code more efficient.)

@kevinbarabash Agreed that pre-commit or pre-push hook would be nice!

xymostech commented:
Yay! Thank you for the cross-browser testing! I have a Safari that I have access to and can test if nobody else does.

+1 to the pre-commmit hooks. I'm always a little sad because I think that when you clone a repository, you can't make it also add some commit hooks by default, so you have to manually set them up. :(

edemaine commented:
Googling around, apparently [shared-git-hooks](https://github.com/kilianc/shared-git-hooks) can enforce Git hooks via `npm install`.

mbourne commented:
Regarding

> when selecting and copying whole KaTeX-rendered elements

My experience using Chrome/Windows of the fiddle mentioned earlier:

If I copy the display math portion only I get this:

> x^2+y^2=z^2 x 
> 2
>  +y 
> 2
>  =z 
> 2

If I include just a bit of text above and below the block, I get this:

> math:
>  $$ x^2+y^2=z^2 $$
> The

I suspect most users will copy just the math, and will not get the desired result.

kevinbarabash commented:
I ran into the same issue.  Maybe we can walk up the tree until we hit a `.katex` node so that it's a little more forgiving with the selection.

edemaine commented:
I thought it might be useful to allow people to copy/paste part of an expression (e.g. a `\text` chunk), and in that case we can't get the source... but perhaps there's a situation in which you select an entire math block but it doesn't give us the div.  I'll do some more experimentation.  We could use `user-select: all` to force all-or-nothing selection behavior, though that might also be annoying to the user...

edemaine commented:
OK, when selecting exactly a full formula, you don't get the containing `.katex` div (even though it's entirely selected!).  I've separated the handling of `.katex-mathml` and `.katex-html` so that we don't need the containing `.katex` div.  So now, if you fully select the display formula in Chrome, you'll get `$x^2+y^2=z^2 $`.  I can't detect the `$$` because I don't have the containing `.katex-display` either...  Also, if you select by clicking on the x and then dragging right, you don't end up selecting the MathML part at all, so you're back to selecting a subset of the formula.  Ouch, and this works only in Chrome, not Firefox.

At this point, I've gone as far as I can with this approach, `window.getSelection.getRangeAt(0).cloneContents()`.  There are two alternatives:

1. Use a `TreeWalker`.  See [this guide](https://superdevelopment.com/2016/02/08/working-with-the-html-selection-api/) which I've been following.  I'm not quite sure what behavior to target, though.
2. Define `user-select: all` on `.katex` and `.katex-display`.  Presumably just within this `contrib` package.  I've done some testing and this seems to work pretty well.  Just might annoy some users trying to select a piece of a big formula.

Probably option 2 is best, but let me know your thoughts.  Of course, a developer can always choose to use the JavaScript and not this CSS, and deal with the consequences.

kevinbarabash commented:
Safari's `select-all` behavior isn't very good.  It copies the right thing, but what it shows for its selection highlight is ridiculous.
![screen shot 2017-08-26 at 10 16 04 pm](https://user-images.githubusercontent.com/1044413/29746564-33d9e542-8aac-11e7-9296-cdfd71772408.png)

kevinbarabash commented:
<img width="269" alt="screen shot 2017-08-26 at 10 22 40 pm" src="https://user-images.githubusercontent.com/1044413/29746588-1f477e68-8aad-11e7-805d-b74e3787dcae.png">

The `KaTeX` logo selection works fine.  My guess is probably some issue with vertical layouts.

kevinbarabash commented on contrib/copy-tex/README.md:
```diff
@@ -0,0 +1,26 @@
+[Clipboard API](https://developer.mozilla.org/en-US/docs/Web/API/ClipboardEvent)
+so that, when selecting and copying whole KaTeX-rendered elements, the text
+content of the resulting clipboard renders KaTeX elements as their LaTeX source
+surrounded by specified delimiters.  (The HTML content remains unchanged.)
```
Maybe: "The rendered HTML content on the page remains unchanged."

edemaine replied:
Actually, I meant that the HTML content of the copy/paste buffer remains unchanged.  Will revise.

kevinbarabash commented on contrib/copy-tex/copy-tex.js:
```diff
@@ -0,0 +1,71 @@
+
+// Replace .katex elements with their TeX source (<annotation> element).
+// Modifies fragment in-place.  Useful for writing your own 'copy' handler.
+window.katexReplaceWithTex = function(fragment) {
```
`katexReplaceWithTex` doesn't need to go on `window`.  In fact this whole file could be wrapped in an IIFE to avoid polluting the global namespace.

edemaine replied:
The whole thing is already wrapped in an IIFE because it gets babelified into `build/contrib/copy-tex*.js`.  I was polluting the namespace in case someone wanted to build their own copy handler using this function.  In hindsight, that's not a very nice way to do it.

Instead I've split the function into its own module, `katex2tex.js`, which is the re-usable part.  (This is similar to how `auto-render` does it.)  Babelify merges these two files into one.  No more namespace pollution.

You still can directly change the delimiters only by manually editing the build.  (This is necessary if we don't want to expose any pollution.)  But now you could copy/paste the rather short `copy-tex.js` into your own build environment (and `require('katex/contrib/copy-tex/katex2tex.js')`).

kevinbarabash replied:
right... I forgot we started babelifying the contrib folder.

edemaine commented on Makefile:
```diff
@@ -48,7 +49,7 @@ build/katex.css: static/katex.less $(wildcard static/*.less) $(NIS)
 
 build/katex.min.css: build/katex.css
-	./node_modules/.bin/cleancss -o $@ $<
+	$(CLEANCSS) -o $@ $<
```
Exactly.  Just assigning it a variable so that it gets passed down (with necessary `..`s) into the subdirectory makes, like the other variables.

edemaine commented:
New commits should deal with all requested changes, in particular in documentation.  Also dealt with a small conflict on `Makefile` by merging master in.
